Fixes cluster state-based role mappings not recovered from disk

albertzaharovits opened the pull request:
Cluster state role mappings need to be recovered from disk CS
(when master nodes are restarted), because the CS metadata
that says if the role mappings are up to date (it actually tells
if all file settings are up to date) is itself recovered from disk.
In case that the metadata is recovered, but the role mappings
themselves are not, the role mappings from the file will not
be present in the CS and will also not be copied from the
settings file.

Fixes: https://github.com/elastic/elasticsearch/issues/93048

albertzaharovits committed:
Parse XContentRoleMappings

albertzaharovits committed:
XContentParsing is WRONG

albertzaharovits committed:
Tests fixed

albertzaharovits committed:
Merge branch 'main' into parse-role-mappings-from-persisted-cluster-state

jonathan-buttner committed:
[ML] Ignoring duplicate keys for _cat trained models and usage (#109126)

* Handling duplicate keys

* Update docs/changelog/109126.yaml

* adding tests

---------

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

albertzaharovits committed:
Expanded tests some more

albertzaharovits commented on server/src/internalClusterTest/java/org/elasticsearch/reservedstate/service/FileSettingsServiceIT.java:
```diff
@@ -8,6 +8,7 @@
 
```
changes in this file are just to expose some methods to reuse in another test

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackClientPlugin.java:
```diff
@@ -376,10 +376,16 @@ public List<NamedXContentRegistry.Entry> getNamedXContent() {
                 PersistentTaskParams.class,
                 new ParseField(SecurityMigrationTaskParams.TASK_NAME),
                 SecurityMigrationTaskParams::fromXContent
+            ),
+            new NamedXContentRegistry.Entry(
+                Metadata.Custom.class,
+                new ParseField(RoleMappingMetadata.TYPE),
+                RoleMappingMetadata::fromXContent
```
This is the essential part of this PR.

albertzaharovits commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/FileSettingsRoleMappingsRestartIT.java:
```diff
@@ -8,33 +8,26 @@
 package org.elasticsearch.xpack.security;
 
 import org.elasticsearch.action.admin.cluster.state.ClusterStateRequest;
-import org.elasticsearch.cluster.ClusterChangedEvent;
```
This is the IT that exercises that role mappings are recovered from disk CS.

albertzaharovits committed:
Update docs/changelog/109167.yaml

elasticsearchmachine commented:
Hi @albertzaharovits, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

nik9000 committed:
ESQL: Make `NameId` `Writeable` (#109164)

This exposes the read and write logic for `NameId` to ESQL-core so we
can implement `Writeable` on more things to make ESQL more like the rest
of Elasticsearch. Reading a `NameId` has to go through the resolution
logic so you still need to read it via the `PlanStreamInput`. `NameId`
is just special.

jakelandis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleMappingMetadata.java:
```diff
@@ -71,9 +95,24 @@ public static NamedDiff<Metadata.Custom> readDiffFrom(StreamInput streamInput) t
 
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        return ChunkedToXContent.wrapAsToXContent(this).toXContent(builder, params);
```
can you help me understand this change ? I am mostly concerned if this will change the format that is persisted to cluster state in any way. I suspect we want to avoid 2 variants of persisted state.

albertzaharovits replied:
Thanks for raising this! It's actually **not** required, and I've removed it.
I needed it only for tests, i.e. because `RoleMappingMetadataXContentSerializationTests` extended `AbstractXContentSerializingTestCase`. But I then found out about `AbstractChunkedSerializingTestCase` so now `RoleMappingMetadata` doesn't need to implement anything new (`ToXContent`).

jakelandis commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleMappingMetadata.java:
```diff
@@ -16,22 +16,46 @@
+        PARSER.declareObjectArray(
+            constructorArg(),
+            // role mapping names are lost when the role mapping metadata is serialized
+            (p, c) -> ExpressionRoleMapping.parse("name_not_available_after_deserialization", p),
```
This further tightly couples the payload of our REST API to the payload of cluster state, despite the workflow being disjoint. I think this is fine for now, but it does present some long term concern w.r.t. these objects and the tight coupling of their serialization. The mismatch between persisted state across multiple versions of ES vs. transient state in a REST API yields different expectations and want to protect against good nature changes in the REST API breaking cluster state and vice-versa. Given the impact of getting this wrong at some future point, can you follow up with something that decouples these two ? Perhaps an adaptor pattern, where today it is an identity adaption (e.g. no differences) but is explicit and tests that changes to one will not influence the other.

albertzaharovits replied:
You're right to point out that this is an issue. I'll follow-up with an implementation of "`ExpressionRoleMapping` in the CS" separate from "`ExpressionRoleMapping` in request payload".

jakelandis reviewed (commented):
looking good ... (reviewing tests now)

jakelandis reviewed (approved):
LGTM assuming https://github.com/elastic/elasticsearch/pull/109167/files#r1619371128 does not introduce a new variant of persistent meta data in cluster state.

arteam committed:
AwaitsFix: https://github.com/elastic/elasticsearch/issues/109188

arteam committed:
AwaitsFix: https://github.com/elastic/elasticsearch/issues/109189

arteam committed:
AwaitsFix: https://github.com/elastic/elasticsearch/issues/109190

rjernst committed:
Guard against a null scorer in painless execute (#109048)

The painless execute api allows for a sample document and query to be
provided. If the query does not actually match the document, a null
scorer is produced. This commit guards against that condition, returning
a clearer error message indicating what happened.

closes #43541

jfreden committed:
AwaitsFix: https://github.com/elastic/elasticsearch/issues/109198

albertzaharovits committed:
Merge branch 'main' into parse-role-mappings-from-persisted-cluster-state

albertzaharovits committed:
Drop "implements ToXContent"

albertzaharovits committed:
nit

albertzaharovits committed:
nit2

albertzaharovits commented:
I really appreciate the lighting review on this one Jake!

albertzaharovits commented:
@gwbrown I rushed with merging this one given its potential impact. I'm open to discuss if you have comments on the approach.

gwbrown commented:
👍🏼 100% behind merging with just Jake's review, I don't have any comments beyond what he left.
